### PR TITLE
sectionBy macro

### PR DESCRIPTION
**Pull request by AlexVanderbist**

### CODE REVIEW

File: src/macros.php [hunk 1]
```diff
@@ -189,6 +189,46 @@
...
+            $currentKey = key($results) ?? 0;
+
+            if ($previousSectionKey && $previousSectionKey != $sectionKey) {
+                $currentKey++;
+            }
+
+            $results[$currentKey][$sectionByKey] = $sectionKey;
+
+            if (! array_key_exists('items', $results[$currentKey])) {
+                $results[$currentKey]['items'] = new static;
```
Comment (sebastiandedeyne) [inline src/macros.php]: Always use braces when instantiating an object

File: src/macros.php [hunk 2]
```diff
@@ -189,6 +189,46 @@
...
 
+if (! Collection::hasMacro('sectionBy')) {
+    /*
+     * Splits a collection into sections grouped by a given key.
+     *
+     * @param string $sectionBy
+     *
+     * @return \Illuminate\Support\Collection
+     */
+    Collection::macro('sectionBy', function ($sectionByKey) {
```
Comment (freekmurze) [inline src/macros.php]: This logic feels somewhat heavy and unreadable to me.

Could we accomplish the same results by using a `groupBy` first and than a `map` to get the desired format?

Comment (sebastiandedeyne) [inline src/macros.php]: `groupBy` doesn't allow duplicate keys

File: src/macros.php [hunk 3]
```diff
@@ -200,30 +200,22 @@
...
 
         foreach ($this->items as $key => $value) {
             $sectionKey = $sectionKeyRetriever($value);
 
-            end($results);
-            $currentKey = key($results) ?? 0;
-
-            if ($previousSectionKey && $previousSectionKey != $sectionKey) {
-                $currentKey++;
+            if (! $results->last() || $results->last()->get($sectionByKey) != $sectionKey) {
```
Comment (sebastiandedeyne) [inline src/macros.php]: Use a strict comparison here if possible

**Comment by sebastiandedeyne** [comment]
Released in `2.4.0`!

**Pull request merged**